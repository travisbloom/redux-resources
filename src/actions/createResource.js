import { capitalize, generateConstantFromString } from '../utilities'
import {
    CREATE_RESOURCE_REQUEST,
    CREATE_RESOURCE_REQUEST_SUCCESS,
    CREATE_RESOURCE_REQUEST_ERROR,
} from '../constants'

const createResource = ({ formatErrors, request, resource, normalizer }) => {
    const resourceActionTypeName = generateConstantFromString(resource)
    const initialActionType = `CREATE_${resourceActionTypeName}_REQUEST`
    const successActionType = `${initialActionType}_SUCCESS`
    const errorActionType = `${initialActionType}_ERROR`
    const actionCreatorName = `create${capitalize(resource)}`

    const createResourceActionCreator = (body, options = {}) => (dispatch, getState) => {
        const { shouldThrowErrors } = options
        const meta = { ...options, timestamp: new Date().toISOString(), resource }
        dispatch({
            type: initialActionType,
            payload: body,
            meta: { ...meta, reduxResourcesActionType: CREATE_RESOURCE_REQUEST },
        })
        return request(body, { ...options, dispatch, getState }).then((response) => {
            dispatch({
                type: successActionType,
                payload: response,
                meta: {
                    ...meta,
                    normalizedResponse: normalizer(response),
                    reduxResourcesActionType: CREATE_RESOURCE_REQUEST_SUCCESS,
                },
            })
            return response
        })
        .catch((response) => {
            dispatch({
                type: errorActionType,
                payload: formatErrors(response),
                meta: { ...meta, reduxResourcesActionType: CREATE_RESOURCE_REQUEST_ERROR },
            })
            if (shouldThrowErrors) throw response
        })
    }
    return {
        [initialActionType]: initialActionType,
        [successActionType]: successActionType,
        [errorActionType]: errorActionType,
        [actionCreatorName]: createResourceActionCreator,
    }
}

export default createResource
