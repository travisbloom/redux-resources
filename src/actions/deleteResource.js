import { capitalize, generateConstantFromString } from '../utilities'
import {
    DELETE_RESOURCE_REQUEST,
    DELETE_RESOURCE_REQUEST_SUCCESS,
    DELETE_RESOURCE_REQUEST_ERROR,
} from '../constants'

const deleteResource = ({ formatErrors, request, resource }) => {
    const resourceActionTypeName = generateConstantFromString(resource)
    const initialActionType = `DELETE_${resourceActionTypeName}_REQUEST`
    const successActionType = `${initialActionType}_SUCCESS`
    const errorActionType = `${initialActionType}_ERROR`
    const actionCreatorName = `delete${capitalize(resource)}`

    const deleteResourceActionCreator = (id, options = {}) => (dispatch, getState) => {
        const { shouldThrowErrors } = options
        const meta = { ...options, resource, timestamp: new Date().toISOString(), id }
        dispatch({
            type: initialActionType,
            payload: id,
            meta: { ...meta, reduxResourcesActionType: DELETE_RESOURCE_REQUEST },
        })
        return request(id, { ...options, dispatch, getState })
            .then(() => {
                dispatch({
                    type: successActionType,
                    payload: id,
                    meta: { ...meta, reduxResourcesActionType: DELETE_RESOURCE_REQUEST_SUCCESS },
                })
            })
            .catch((response) => {
                dispatch({
                    type: errorActionType,
                    payload: formatErrors(response),
                    meta: { ...meta, reduxResourcesActionType: DELETE_RESOURCE_REQUEST_ERROR },
                })
                if (shouldThrowErrors) throw response
            })
    }
    return {
        [initialActionType]: initialActionType,
        [successActionType]: successActionType,
        [errorActionType]: errorActionType,
        [actionCreatorName]: deleteResourceActionCreator,
    }
}

export default deleteResource
