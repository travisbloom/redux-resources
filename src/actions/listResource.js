import { serializeOrderedQuery, capitalize, generateConstantFromString } from '../utilities'
import {
    selectResourceList,
    selectSelectedResourceList,
    selectResourceListsBeingFetched,
} from '../selectors'
import {
    LIST_RESOURCE_REQUEST,
    LIST_RESOURCE_REQUEST_SUCCESS,
    LIST_RESOURCE_REQUEST_ERROR,
    SELECT_RESOURCE_LIST,
    DEFAULT_QUERY_LIMIT,
} from '../constants'

const defaultIsCached = obj => !!obj

const listResource = ({
    formatErrors,
    resource,
    request,
    normalizer,
    isCachedFn,
    stringifyQueryFn,
}) => {
    const resourceActionTypeName = generateConstantFromString(resource)
    const initialActionType = `LIST_${resourceActionTypeName}_REQUEST`
    const successActionType = `${initialActionType}_SUCCESS`
    const errorActionType = `${initialActionType}_ERROR`
    const selectActionType = `SELECT_${resourceActionTypeName}_LIST`
    const actionCreatorName = `list${capitalize(resource)}`

    const isCached = isCachedFn || defaultIsCached
    const stringifyQuery = stringifyQueryFn || serializeOrderedQuery
    /**
    * @param {object} the query param object being used to filter the resource request
    * @param {object} the options object
    *     @property {boolean} shouldIgnoreCache - if true, will fetch the remote resource even if it
          has been previously retrieved
    *     @property {boolean} shouldThrowErrors - if true, will not pass returned errors to redux
          but instead throw them so the fn calling can .catch them
    */
    const listResourceActionCreator = (passedQuery = {}, options = {}) => (dispatch, getState) => {
        const { shouldThrowErrors, shouldIgnoreCache } = options
        const query = {
            limit: DEFAULT_QUERY_LIMIT,
            offset: 0,
            ...passedQuery,
        }

        const queryString = stringifyQuery(query)
        const state = getState()
        const previousResourceList = selectResourceList(state, resource, queryString)
        const meta = {
            query,
            queryString,
            resource,
            timestamp: new Date().toISOString(),
            ...options,
        }

        // if the resource already exists in redux
        if (!shouldIgnoreCache && isCached(previousResourceList, query, options)) {
            // if the requested resource is not the current selected resource
            if (selectSelectedResourceList(state, resource) !== queryString) {
                dispatch({
                    type: selectActionType,
                    payload: queryString,
                    meta: { ...meta, reduxResourcesActionType: SELECT_RESOURCE_LIST },
                })
            }
            return Promise.resolve()
        }

        // if the resource is currently being fetched already
        if (selectResourceListsBeingFetched(state, resource).includes(queryString)) {
            return Promise.resolve()
        }

        dispatch({
            type: initialActionType,
            payload: queryString,
            meta: { ...meta, reduxResourcesActionType: LIST_RESOURCE_REQUEST },
        })

        return request(query, { ...options, dispatch, getState }).then((response) => {
            dispatch({
                type: successActionType,
                payload: response,
                meta: {
                    ...meta,
                    normalizedResponse: normalizer(response),
                    reduxResourcesActionType: LIST_RESOURCE_REQUEST_SUCCESS,
                },
            })
        })
        .catch((response) => {
            dispatch({
                type: errorActionType,
                payload: formatErrors(response),
                meta: {
                    ...meta,
                    reduxResourcesActionType: LIST_RESOURCE_REQUEST_ERROR,
                },
            })
            if (shouldThrowErrors) throw response
        })
    }
    return {
        [initialActionType]: initialActionType,
        [successActionType]: successActionType,
        [errorActionType]: errorActionType,
        [selectActionType]: selectActionType,
        [actionCreatorName]: listResourceActionCreator,
    }
}

export default listResource
