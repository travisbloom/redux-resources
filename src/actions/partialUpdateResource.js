import { capitalize, generateConstantFromString } from '../utilities'
import {
    selectResource,
} from '../selectors'
import {
    PARTIAL_UPDATE_RESOURCE_REQUEST,
    PARTIAL_UPDATE_RESOURCE_REQUEST_SUCCESS,
    PARTIAL_UPDATE_RESOURCE_REQUEST_ERROR,
} from '../constants'

/**
* update a resource, either remotely or just locally.
* @param {object} the config object for getResource
*     @property {string} resource - see above for details
*     @property {function} request - see above for details
*     @property {function} normalizer - see above for details
*     @property {string} resourceActionName - see above for details
*/
const partialUpdateResource = ({ request, resource, normalizer, formatErrors }) => {
    const resourceActionTypeName = generateConstantFromString(resource)
    const initialActionType = `PARTIAL_UPDATE_${resourceActionTypeName}_REQUEST`
    const successActionType = `${initialActionType}_SUCCESS`
    const errorActionType = `${initialActionType}_ERROR`
    const actionCreatorName = `partialUpdate${capitalize(resource)}`
        /**
        * @param {string|number} the ID of the resource being updated
        * @param {object} the json body payload of the update
        * @param {object} the options object
        *     @property {boolean} shouldThrowErrors - if true, will not pass returned
              errors to redux but instead throw them so the fn calling can .catch them
        *     @property {boolean} isOptimisticUpdate: determines whether the new fields
              should update before or after the server request
              (this will revert the changes on error)
        **/

    const partialUpdateResourceActionCreator = (id, body, options = {}) => (dispatch, getState) => {
        const { isOptimisticUpdate, shouldThrowErrors } = options
        const meta = { ...options, resource, id, timestamp: new Date().toISOString() }
        if (isOptimisticUpdate) {
            const state = getState()
                // we store the replaced resource here so we can revert if it fails
            const replacedResource = selectResource(state, resource, id)
            const replacedLastUpdatedAt = state.resources[resource].lastUpdatedAt

            const updatedResource = { ...replacedResource, ...body }
            dispatch({
                type: successActionType,
                payload: updatedResource,
                meta: {
                    ...meta,
                    normalizedResponse: normalizer(updatedResource),
                    reduxResourcesActionType: PARTIAL_UPDATE_RESOURCE_REQUEST_SUCCESS,
                },
            })

            return request(id, body, { ...options, dispatch, getState }).catch((response) => {
                dispatch({
                    type: errorActionType,
                    payload: formatErrors(response),
                    meta: {
                        ...meta,
                        replacedLastUpdatedAt,
                        replacedResource: { ...replacedResource },
                        reduxResourcesActionType: PARTIAL_UPDATE_RESOURCE_REQUEST_ERROR,
                    },
                })
                if (shouldThrowErrors) throw response
            })
        }

        dispatch({
            type: initialActionType,
            payload: id,
            meta: { ...meta, reduxResourcesActionType: PARTIAL_UPDATE_RESOURCE_REQUEST },
        })

        return request(id, body, { ...options, dispatch, getState }).then((response) => {
            dispatch({
                type: successActionType,
                payload: response,
                meta: {
                    ...meta,
                    normalizedResponse: normalizer(response),
                    reduxResourcesActionType: PARTIAL_UPDATE_RESOURCE_REQUEST_SUCCESS,
                },
            })
        })
        .catch((response) => {
            dispatch({
                type: errorActionType,
                payload: formatErrors(response),
                meta: {
                    ...meta,
                    reduxResourcesActionType: PARTIAL_UPDATE_RESOURCE_REQUEST_ERROR,
                },
            })
            if (shouldThrowErrors) throw response
        })
    }

    return {
        [initialActionType]: initialActionType,
        [successActionType]: successActionType,
        [errorActionType]: errorActionType,
        [actionCreatorName]: partialUpdateResourceActionCreator,
    }
}

export default partialUpdateResource
