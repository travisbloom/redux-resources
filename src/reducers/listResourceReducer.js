import {
    LIST_RESOURCE_REQUEST,
    LIST_RESOURCE_REQUEST_SUCCESS,
    LIST_RESOURCE_REQUEST_ERROR,
    SELECT_RESOURCE_LIST,
} from '../constants'
import { removeFromArray, addToArray } from '../utilities'

const listResourceReducer = (state, action) => {
    const { payload, meta } = action
    const {
        query,
        queryString,
        resource,
        timestamp,
        reduxResourcesActionType,
        normalizedResponse,
    } = meta
    const resourceState = state[resource]

    switch (reduxResourcesActionType) {
    case SELECT_RESOURCE_LIST:
        return {
            ...state,
            [resource]: {
                ...resourceState,
                selectedResourceList: queryString,
            },
        }
    case LIST_RESOURCE_REQUEST:
        return {
            ...state,
            [resource]: {
                ...resourceState,
                isFetching: true,
                resourceListsBeingFetched: addToArray(
                    queryString,
                    resourceState.resourceListsBeingFetched
                ),
            },
        }
    case LIST_RESOURCE_REQUEST_ERROR:
        return {
            ...state,
            [resource]: {
                ...resourceState,
                isFetching: false,
                resourceListsBeingFetched: removeFromArray(
                    queryString,
                    resourceState.resourceListsBeingFetched
                ),
                errors: [
                    ...resourceState.errors,
                    ...payload,
                ],
            },
        }
    case LIST_RESOURCE_REQUEST_SUCCESS: {
        const { entities, result } = normalizedResponse
        return Object.keys(state).reduce((newState, resourceKey) => {
            if (resourceKey !== resource && entities[resourceKey]) {
                newState[resourceKey] = {
                    ...state[resourceKey],
                    resources: {
                        ...state[resourceKey].resources,
                        ...entities[resourceKey],
                    },
                }
            } else if (resourceKey === resource) {
                newState[resourceKey] = {
                    ...state[resourceKey],
                    isFetching: false,
                    resources: {
                        ...state[resourceKey].resources,
                        ...(entities[resource] || {}),
                    },
                    errors: [],
                    resourceListsBeingFetched: removeFromArray(
                        queryString,
                        state[resourceKey].resourceListsBeingFetched
                    ),
                    selectedResourceList: queryString,
                    lastListedAt: timestamp,
                    resourceLists: {
                        ...state[resourceKey].resourceLists,
                        [queryString]: {
                            result,
                            query,
                            queryString,
                        },
                    },
                }
            } else {
                newState[resourceKey] = state[resourceKey]
            }
            return newState
        }, {})
    }
    }
}

export default listResourceReducer
