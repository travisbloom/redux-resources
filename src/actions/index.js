import createResource from './createResource'
import deleteResource from './deleteResource'
import listResource from './listResource'
import retrieveResource from './retrieveResource'
import updateResource from './updateResource'

module.exports = {
    createResource,
    listResource,
    retrieveResource,
    updateResource,
    deleteResource,
}
