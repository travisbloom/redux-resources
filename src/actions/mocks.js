import {
    mockStore,
    normalizer,
    RESOURCE_REDUCER,
    RESOURCE,
    RESOURCE_ID,
    INITIAL_STATE,
    INITIAL_STATE_WITH_CACHED_RESOURCE,
    QUERY,
    RESOURCE_UPDATE_PAYLOAD,
    formatErrors,
} from '../mocks'
import createResource from './createResource'
import deleteResource from './deleteResource'
import listResource from './listResource'
import partialUpdateResource from './partialUpdateResource'
import updateResource from './updateResource'
import retrieveResource from './retrieveResource'

const generateCreateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState)
    const { createMockResources } = createResource({
        resource: RESOURCE_REDUCER,
        normalizer,
        request,
        formatErrors,
    })
    await store.dispatch(createMockResources(RESOURCE, options))
    return store.getActions()
}

const generateDeleteResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { deleteMockResources } = deleteResource({
        resource: RESOURCE_REDUCER,
        request,
        formatErrors,
    })
    await store.dispatch(deleteMockResources(RESOURCE_ID, options))
    return store.getActions()
}

const generateListResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { listMockResources } = listResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(listMockResources(QUERY, options))
    return store.getActions()
}

const generateRetrieveResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { retrieveMockResources } = retrieveResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(retrieveMockResources(RESOURCE_ID, options))
    return store.getActions()
}

const generatePartialUpdateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE_WITH_CACHED_RESOURCE)
    const { partialUpdateMockResources } = partialUpdateResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(partialUpdateMockResources(RESOURCE_ID, RESOURCE_UPDATE_PAYLOAD, options))
    return store.getActions()
}

const generateUpdateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE_WITH_CACHED_RESOURCE)
    const { updateMockResources } = updateResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    const body = { ...RESOURCE, ...RESOURCE_UPDATE_PAYLOAD }
    await store.dispatch(updateMockResources(RESOURCE_ID, body, options))
    return store.getActions()
}

module.exports = {
    generateCreateResourceActions,
    generateDeleteResourceActions,
    generateListResourceActions,
    generateRetrieveResourceActions,
    generatePartialUpdateResourceActions,
    generateUpdateResourceActions,
}
