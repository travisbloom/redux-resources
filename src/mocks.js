import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Schema, normalize, arrayOf } from 'normalizr'

import { getResourceInitialState } from './reducers'

const RESOURCE_REDUCER = 'mockResources'
const RELATED_RESOURCE_REDUCER = 'relatedMockResources'

const RELATED_RESOURCE_ID = 'relatedResourceId'
const RELATED_RESOURCE = {
    id: RELATED_RESOURCE_ID,
    relatedProp1: 'initial relatedProp1',
}
const RESOURCE_ID = 'resourceId'
const RESOURCE = {
    id: RESOURCE_ID,
    prop1: 'initial prop1',
    prop2: 'initial prop2',
    relatedResource: RELATED_RESOURCE,
}
const NORMALIZED_RESOURCE = {
    id: RESOURCE_ID,
    prop1: 'initial prop1',
    prop2: 'initial prop2',
    relatedResource: RELATED_RESOURCE_ID,
}

const RESOURCE_UPDATE_PAYLOAD = {
    id: RESOURCE_ID,
    prop2: 'UPDATED PROP_1',
}

const SERVER_ERRORS = [
    'This is an initial error',
    'This is another error',
]

const QUERY = {
    limit: 30,
    offset: 0,
    someQueryParam: 'someValue',
}
const QUERY_STRING = 'key=30&key=0&key="someValue"'

const INITIAL_STATE = {
    resources: {
        [RELATED_RESOURCE_REDUCER]: getResourceInitialState(),
        [RESOURCE_REDUCER]: getResourceInitialState(),
    },
}
const INITIAL_STATE_WITH_LIST_BEING_FETCHED = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resourceListsBeingFetched: [QUERY_STRING],
        },
    },
}
const INITIAL_STATE_WITH_CACHED_LIST = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resourceLists: {
                [QUERY_STRING]: {
                    result: [RESOURCE_ID],
                },
            },
        },
    },
}
const INITIAL_STATE_WITH_CACHED_AND_SELECTED_LIST = {
    resources: {
        ...INITIAL_STATE_WITH_CACHED_LIST.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE_WITH_CACHED_LIST.resources[RESOURCE_REDUCER],
            selectedResourceList: QUERY_STRING,
        },
    },
}
const INITIAL_STATE_WITH_RESOURCE_BEING_FETCHED = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resourcesBeingFetched: [RESOURCE_ID],
        },
    },
}
const INITIAL_STATE_WITH_CACHED_RESOURCE = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resources: {
                [RESOURCE_ID]: NORMALIZED_RESOURCE,
            },
        },
    },
}
const INITIAL_STATE_WITH_CACHED_AND_SELECTED_RESOURCE = {
    resources: {
        ...INITIAL_STATE_WITH_CACHED_RESOURCE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE_WITH_CACHED_RESOURCE.resources[RESOURCE_REDUCER],
            selectedResource: RESOURCE_ID,
        },
    },
}
const INITIAL_STATE_WITH_CACHED_RESOURCE_AND_CACHED_RESOURCE_LIST = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resources: {
                [RESOURCE_ID]: NORMALIZED_RESOURCE,
            },
            resourceLists: {
                [QUERY_STRING]: {
                    result: [RESOURCE_ID],
                },
            },
        },
    },
}

const resourceSchema = new Schema(RESOURCE_REDUCER)
const relatedResourceSchema = new Schema(RELATED_RESOURCE_REDUCER)
resourceSchema.define({
    relatedResource: relatedResourceSchema,
})
const normalizer = (resources) => {
    const array = Array.isArray(resources) ? resources : [resources]
    return normalize(array, arrayOf(resourceSchema))
}
const formatErrors = errors => errors

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

module.exports = {
    RESOURCE_REDUCER,
    RESOURCE_ID,
    RESOURCE,

    INITIAL_STATE,
    INITIAL_STATE_WITH_LIST_BEING_FETCHED,
    INITIAL_STATE_WITH_CACHED_LIST,
    INITIAL_STATE_WITH_CACHED_AND_SELECTED_LIST,
    INITIAL_STATE_WITH_RESOURCE_BEING_FETCHED,
    INITIAL_STATE_WITH_CACHED_RESOURCE,
    INITIAL_STATE_WITH_CACHED_AND_SELECTED_RESOURCE,
    INITIAL_STATE_WITH_CACHED_RESOURCE_AND_CACHED_RESOURCE_LIST,

    RESOURCE_UPDATE_PAYLOAD,

    SERVER_ERRORS,
    QUERY,
    QUERY_STRING,

    mockStore,
    formatErrors,
    normalizer,
}
