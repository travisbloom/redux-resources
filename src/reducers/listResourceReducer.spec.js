import { RESOURCE, SERVER_ERRORS, INITIAL_STATE } from '../mocks'
import { generateListResourceActions } from '../actions/mocks'

import listResourceReducer from './listResourceReducer'

const initialState = INITIAL_STATE.resources

it('will update the state after the initial action', async () => {
    const [initialAction] = await generateListResourceActions({
        request: () => Promise.resolve([RESOURCE]),
    })
    expect(listResourceReducer(initialState, initialAction)).toMatchSnapshot()
})

it('will update the state after the success action', async () => {
    const [initialAction, successAction] = await generateListResourceActions({
        request: () => Promise.resolve([RESOURCE]),
    })
    const stateAfterInititalAction = listResourceReducer(initialState, initialAction)
    expect(listResourceReducer(stateAfterInititalAction, successAction)).toMatchSnapshot()
})

it('will update the state after the error action', async () => {
    const [initialAction, errorAction] = await generateListResourceActions({
        request: () => Promise.reject(SERVER_ERRORS),
    })
    const stateAfterInititalAction = listResourceReducer(initialState, initialAction)
    expect(listResourceReducer(stateAfterInititalAction, errorAction)).toMatchSnapshot()
})
