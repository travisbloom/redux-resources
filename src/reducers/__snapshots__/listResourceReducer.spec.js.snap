exports[`test will update the state after the error action 1`] = `
Object {
  "mockResources": Object {
    "errors": Array [
      "This is an initial error",
      "This is another error",
    ],
    "isFetching": false,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {},
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
  "relatedMockResources": Object {
    "errors": Array [],
    "isFetching": false,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {},
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
}
`;

exports[`test will update the state after the initial action 1`] = `
Object {
  "mockResources": Object {
    "errors": Array [],
    "isFetching": true,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [
      "key=30&key=0&key=\"someValue\"",
    ],
    "resources": Object {},
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
  "relatedMockResources": Object {
    "errors": Array [],
    "isFetching": false,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {},
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
}
`;

exports[`test will update the state after the success action 1`] = `
Object {
  "mockResources": Object {
    "errors": Array [],
    "isFetching": false,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": "2016-01-01T05:00:00.000Z",
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {
      "key=30&key=0&key=\"someValue\"": Object {
        "query": Object {
          "limit": 30,
          "offset": 0,
          "someQueryParam": "someValue",
        },
        "queryString": "key=30&key=0&key=\"someValue\"",
        "result": Array [
          "resourceId",
        ],
      },
    },
    "resourceListsBeingFetched": Array [],
    "resources": Object {
      "resourceId": Object {
        "id": "resourceId",
        "prop1": "initial prop1",
        "prop2": "initial prop2",
        "relatedResource": "relatedResourceId",
      },
    },
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": "key=30&key=0&key=\"someValue\"",
  },
  "relatedMockResources": Object {
    "errors": Array [],
    "isFetching": false,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {
      "relatedResourceId": Object {
        "id": "relatedResourceId",
        "relatedProp1": "initial relatedProp1",
      },
    },
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
}
`;
